### PR TITLE
Doctrine & Tiecel

### PR DESCRIPTION
**Pull request by jpawlicki**
Implement r5 changes to the Church and Tiecel. Also refactored Character tags into an enum.

### CODE REVIEW

File: backend/src/main/java/com/empire/World.java [hunk 1]
```diff
@@ -1660,19 +1682,44 @@ void gainIncome() {
...
-				if (NationData.getStateReligion(k, World.this).religion != Religion.IRUHAN) getNation(k).goodwill -= 5;
-				if (getNation(k).loyalToCult) getNation(k).goodwill -= 10;
+				Religion religion = stateIdeologies.get(k).religion;
+				if (church.hasDoctrine(Church.Doctrine.ANTIECUMENISM) && religion != Religion.IRUHAN && religion != Religion.NONE) getNation(k).goodwill += Constants.antiecumenismStateOpinion;
+				if (church.hasDoctrine(Church.Doctrine.ANTIAPOSTASY) && getNation(k).loyalToCult) getNation(k).goodwill += Constants.antiapostasyOpinion;
+			}
+			Set<String> iruhanNations = stateIdeologies.entrySet().stream().filter(e -> e.getValue().religion == Religion.IRUHAN).map(e -> e.getKey()).collect(Collectors.toSet());
+			if (church.hasDoctrine(Church.Doctrine.INQUISITION)) {
+				Set<String> vesselNations = stateIdeologies.entrySet().stream().filter(e -> e.getValue() == Ideology.VESSEL_OF_FAITH).map(e -> e.getKey()).collect(Collectors.toSet());
+				for (String k : iruhanNations) {
```
Comment (pawlick2) [inline backend/src/main/java/com/empire/World.java]: If you wanted to go full-on stream mode and replace a bunch of these one-line for-loops you could, as an example say (this looks better when its tabbed nicely):
`iruhanNations.stream().filter(k -> vesselNations.stream().anyMatch(v -> !NationData.isAttackingOnSight(k, v, World.this))).forEach(k -> getNation(k).goodwill += Constants.inquisitionOpinion);`
Admittedly I like to keep the stream nesting to a minimum as a general style (don't cross the streams!) but here I would maybe make an exception.

File: backend/src/main/java/com/empire/World.java [hunk 2]
```diff
@@ -1660,19 +1682,44 @@ void gainIncome() {
...
 		void churchOpinionChangesDueToStateReligion() {
+			Map<String, Ideology> stateIdeologies = kingdoms.keySet().stream().collect(Collectors.toMap(k -> k, k -> NationData.getStateReligion(k, World.this)));
 			for (String k : kingdoms.keySet()) {
-				if (NationData.getStateReligion(k, World.this).religion != Religion.IRUHAN) getNation(k).goodwill -= 5;
-				if (getNation(k).loyalToCult) getNation(k).goodwill -= 10;
+				Religion religion = stateIdeologies.get(k).religion;
+				if (church.hasDoctrine(Church.Doctrine.ANTIECUMENISM) && religion != Religion.IRUHAN && religion != Religion.NONE) getNation(k).goodwill += Constants.antiecumenismStateOpinion;
+				if (church.hasDoctrine(Church.Doctrine.ANTIAPOSTASY) && getNation(k).loyalToCult) getNation(k).goodwill += Constants.antiapostasyOpinion;
+			}
+			Set<String> iruhanNations = stateIdeologies.entrySet().stream().filter(e -> e.getValue().religion == Religion.IRUHAN).map(e -> e.getKey()).collect(Collectors.toSet());
```
Comment (pawlick2) [inline backend/src/main/java/com/empire/World.java]: Could replace these `e -> e.getKey()` with their functional reference `Map.Entry::getKey` if that's your style.

File: backend/src/main/java/com/empire/Character.java [hunk 3]
```diff
@@ -1,12 +1,20 @@
...
+
 import java.util.List;
 import java.util.ArrayList;
 import java.util.Map;
 import java.util.HashMap;
 import java.util.Collections;
 
 class Character {
+	enum Tag {
+		@SerializedName("Cardinal") CARDINAL,
```
[the remark this answers is not in the document]
Comment (jpawlicki) [reply backend/src/main/java/com/empire/Character.java]: Right now I'm standardizing on backward-compatibility, I think we could later take a project-wide look to break compatibility with the records of r2/r3/r4, but don't want to take that step in this PR.

File: backend/src/main/java/com/empire/Church.java [hunk 4]
```diff
@@ -0,0 +1,50 @@
+package com.empire;
+
+import java.util.HashSet;
+import java.util.Set;
+
+public class Church {
+	public enum Doctrine {
+		ANTIAPOSTASY("antiapostasy. Any rulers who have sworn loyalty to the Cult are named apostates and enemies of the Church of Iruhan!", "antiapostacy. The Church follows Iruhan's example of forgiveess by forgiving the sin of aligning with the Cult."),
```
[the remark this answers is not in the document]
Comment (jpawlicki) [reply backend/src/main/java/com/empire/Church.java]: Done

File: backend/src/main/java/com/empire/Church.java [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+		ANTIECUMENISM("antiecumenism. Construction of heathen temples or widespread worship of false gods is condemned by the Church!", "antiecumenism. The Church now seeks to spread the Word of Iruhan in the light of tolerance for other faiths."),
+		ANTISCHISMATICISM("antischismaticism. The Church declares the Vessel of Faith interpretation of Iruhan's words heretical, and condemns the construction of temples with that emphasis!", "antischismaticism. The Church accepts the right to to study Iruhan's teachings without its direct guidance, albeit unwise."),
```
[the remark this answers is not in the document]
Comment (jpawlicki) [reply backend/src/main/java/com/empire/Church.java]: Done.

File: backend/src/main/java/com/empire/Noble.java [hunk 6]
```diff
@@ -102,15 +102,15 @@ boolean isCreateable(World w, Region r, Map<Army, Character> leaders, Map<Region
...
+				for (Character c : w.characters) if (c.hasTag(Character.Tag.RULER) && c.kingdom.equals(r.getKingdom()) && c.location == w.regions.indexOf(r)) return true;
 				return false;
 			}
 
 			@Override
 			boolean isCreateable(World w, Region r, Map<Army, Character> leaders, Map<Region, ArrayList<Character>> governors, Set<Region> builds, Set<Region> templeBuilds, Map<String, Double> rationing, Set<String> lastStands, int inspires) {
 				Set<Integer> closeRegions = r.getCloseRegionIds(w, Constants.nobleCrisisFrequency);
 				for (Character c : w.characters) {
-					if (c.kingdom.equals(r.getKingdom()) && c.captor.equals("") && c.hasTag("Ruler") && closeRegions.contains(c.location)) {
+					if (c.kingdom.equals(r.getKingdom()) && c.captor.equals("") && c.hasTag(Character.Tag.RULER) && closeRegions.contains(c.location)) {
```
[the remark this answers is not in the document]
Comment (jpawlicki) [reply backend/src/main/java/com/empire/Noble.java]: Done, also updated other sites.

File: backend/src/main/java/com/empire/World.java (hunk 2 above)
Comment (jpawlicki) [reply backend/src/main/java/com/empire/World.java]: Done.

File: backend/src/main/java/com/empire/World.java (hunk 1 above)
Comment (jpawlicki) [reply backend/src/main/java/com/empire/World.java]: Rewrote it in full streaming style, using Google style indentations (two indentation levels on a line continuation, box-model code layout). Take a look and let me know what you think.

Overall, I still feel somewhat undecided on streams versus imperative control flow. The functional stuff works perfectly for obvious set transformations, etc, but once you start pulling in anyMatch / allMatch and operating on nested streams I think it starts to look a little pretty lisp-y. Whether that's a good or bad thing probably depends on who you ask...

File: backend/src/main/java/com/empire/World.java (hunk 2 above)
Comment (jpawlicki) [reply backend/src/main/java/com/empire/World.java]: Done.

File: backend/src/main/java/com/empire/World.java (hunk 1 above)
Comment (pawlick2) [reply backend/src/main/java/com/empire/World.java]: Seeing it this way...it looks not good I think, mainly due to the amount of lines they now take up.  I would usually only allow myself to nest streams when the nested one was a quick one-liner.  With this style of formatting I don't think there can be such thing as a one-liner stream.  Hate to say it but if this is the alternative I vote to put back the loops.

File: backend/src/main/java/com/empire/Character.java [hunk 7]
```diff
@@ -81,16 +89,20 @@ public void addExperienceGovernor() {
...
+	void addTag(Tag tag) {
 		tags.add(tag);
 	}
 
-	void removeTag(String tag) {
+	void removeTag(Tag tag) {
 		tags.remove(tag);
 	}
+
+	boolean captive() {
```
Comment (pawlick2) [inline backend/src/main/java/com/empire/Character.java]: One final thing, I think this should be made to isCaptive() so the resulting code reads more like a sentence.

File: backend/src/main/java/com/empire/World.java (hunk 1 above)
Comment (jpawlicki) [reply backend/src/main/java/com/empire/World.java]: One alternative that deals with the indentation but not the line count is to eliminate the control flow (the loops) but keep using variables (e.g. vesselNations, or even could define the filter predicate itself as a variable). Still, the operation is complex.

Comment (pawlick2) [reply backend/src/main/java/com/empire/World.java]: Yes, in my original example I was keeping the variables you had already defined.  Personally, I love using streams, however I've shot myself in the foot enough times by overdoing it with them.  The medium approach using variables would be acceptable to me, doing each of the ops in a single statement I think is too much.

File: backend/src/main/java/com/empire/Character.java (hunk 7 above)
Comment (jpawlicki) [reply backend/src/main/java/com/empire/Character.java]: Done.

File: backend/src/main/java/com/empire/World.java (hunk 1 above)
Comment (jpawlicki) [reply backend/src/main/java/com/empire/World.java]: Named the predicates - I think that's a bit clearer, if more verbose.

**Pull request merged**